renderdoc/driver/gl/gl_pixelhistory.cpp

```diff
@@ -422,27 +717,75 @@ std::map<uint32_t, uint32_t> QueryNumFragmentsByEvent(WrappedOpenGL *driver,
     driver->glStencilMask(0xff);    // default for 1 byte
     driver->glStencilFunc(eGL_ALWAYS, 0, 0xff);
     driver->glClearStencil(0);
-    driver->glClear(eGL_STENCIL_BUFFER_BIT);
+    driver->glClearColor(0, 0, 0, 0);
+    driver->glClearDepth(0);
+    driver->glClear(eGL_STENCIL_BUFFER_BIT | eGL_COLOR_BUFFER_BIT | eGL_DEPTH_BUFFER_BIT);
     driver->glEnable(eGL_STENCIL_TEST);
     // depth test enable
     driver->glEnable(eGL_DEPTH_TEST);
     driver->glDepthFunc(eGL_ALWAYS);
     driver->glDepthMask(GL_TRUE);
     driver->glDisable(eGL_BLEND);
 
+    // enable the sample we're looking at so we get the shaderOut even if it's masked off
+    driver->glEnable(eGL_SAMPLE_MASK);
+    if(sampleIndex != ~0u)
+    {
+      driver->glSampleMaski(0, 1u << sampleIndex);
+    }
+    else
+    {
+      driver->glSampleMaski(0, ~0u);
+    }
+
     // replay event
     driver->ReplayLog(modEvents[i].eventId, modEvents[i].eventId, eReplay_OnlyDraw);
 
-    ModificationValue modValue;
-    driver->glReadPixels(x, y, 1, 1, eGL_RGBA, eGL_FLOAT, (void *)modValue.col.floatValue.data());
-    driver->glReadPixels(x, y, 1, 1, eGL_DEPTH_COMPONENT, eGL_FLOAT, (void *)&modValue.depth);
     uint32_t numFragments = 0;
-    driver->glReadPixels(x, y, 1, 1, eGL_STENCIL_INDEX, eGL_UNSIGNED_INT, (void *)&numFragments);
-    // We're not reading the stencil value here, so use the postMod instead.
-    // Shaders don't actually output stencil values, those are determined by the stencil op.
-    modValue.stencil = history[i].postMod.stencil;
+    if(sampleIndex == ~0u)
+    {
+      ModificationValue modValue;
+      // if multisampled, then we need to blit to a nonMS texture to read the pixels
```

Comment:
If I'm following correctly, this is doing a resolve when sampleIndex is `~0U` and retrieving the resolved colour for shader output and modification values?

I'd need to double check but I believe the behaviour should be in that case to default to sample 0 for multisampled textures. We should always deal with a concrete output - if the user has 'average value' selected and they run the pixel history then picking 0 gives a more intuitive result I feel, though it's maybe something that should be discouraged at the UI level.

Reply:
Yes, we were using the resolved colour. We can default it to use sample 0 when average value is selected.